### DISCUSSION TITLE
How to run trame + paraview web app on Windows

### DISCUSSION THREAD

**Discussion by mashirafarid** [Kitware/trame#582] [Trame/ParaView]
Hi there, I'm developing a web app using Trame and Paraview. I've been using MacOS and following the instructions [here](https://github.com/Kitware/paraview-visualizer) to run my web app with no problems. I now want to try installing and running my web app on Windows, however I'm having a bit of trouble following the same instructions as before to run the app.

I've tried running the following commands to get the app running:
```powershell
set PVPYTHON="C:\Program Files\ParaView 5.12.1\bin\pvpython"
set PV_VENV=$PWD/.venv
set TRAME_APP=mitraview.app
$PVPYTHON -m paraview.apps.trame
``` 

However, I get the following error:
```powershell
At line:1 char:11
+ $PVPYTHON -m paraview.apps.trame
+           ~~
Unexpected token '-m' in expression or statement.
At line:1 char:14
+ $PVPYTHON -m paraview.apps.trame
+              ~~~~~~~~~~~~~~~~~~~
Unexpected token 'paraview.apps.trame' in expression or statement.
    + CategoryInfo          : ParserError: (:) [], ParentContainsErrorRecordException
    + FullyQualifiedErrorId : UnexpectedToken
``` 

If I remove the quotation marks around the pvpython path, I get the following error:
```powershell
> set PVPYTHON=C:\Program Files\ParaView 5.12.1\bin\pvpython
Set-Variable : A positional parameter cannot be found that accepts argument '5.12.1\bin\pvpython'.
At line:1 char:1
+ set PVPYTHON=C:\Program Files\ParaView 5.12.1\bin\pvpython
+ ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : InvalidArgument: (:) [Set-Variable], ParameterBindingException
    + FullyQualifiedErrorId : PositionalParameterNotFound,Microsoft.PowerShell.Commands.SetVariableCommand
``` 

This is my first time trying to develop on a WIndows machine, so I have no idea what the problem is. Any help would be appreciated!

**Comment by jourdain** [comment]
on windows the executable is named `.../pvpython.exe`

**Reply by mashirafarid** [reply]
Unfortunately, I still get the same errors after using `pvpython.exe`

**Reply by mashirafarid** [reply]
After playing around with different commands, I found that this works:
`& "C:\Program Files\ParaView 5.12.1\bin\pvpython" -m paraview.apps.trame --trame-app mitraview.app --venv .venv`
However, to make things simpler, I'd like to use similar commands to the one I use on MacOS i.e. use environment variables, is this possible?

**Reply by jourdain** [reply]
Yes you can still use the ENV variables.

**Reply by mashirafarid** [reply]
Thanks! I had another question regarding writing tests for this web app. I'm using pytest to write some tests, however when I run them within the venv I get this error: `ModuleNotFoundError: No module named 'paraview'`.

Looking at how [Paraview Visualizer](https://github.com/Kitware/paraview-visualizer/blob/53300065d8edf349625f2e030719faf3eb023439/tests/test_import.py) handles testing, it seems to ignore this error. Can you provide me with any guidance on how to write tests for web apps that use both Trame and Paraview?

**Reply by jourdain** [reply]
You will need to run pytest with pvpython or setup some env variables to enable ParaView within your venv.

**Reply by mashirafarid** [reply]
Sorry, how would I do either of these? I have pytest installed in my venv and I think I tried to run pytest with pvpython, but I get the following error both inside and outside the venv:
```shell
(.venv)% $PVPYTHON -m pytest --venv .venv
/Applications/ParaView-5.12.1.app/Contents/Libraries/vtkpython: No module named pytest
``` 
I also tried this command, which gives me a different error:
```shell
(.venv)% $PVPYTHON tests/test_drawer.py -m pytest             
Traceback (most recent call last):
  File "/Users/mashirafarid/Documents/mitraview/tests/test_drawer.py", line 3, in <module>
    from mitraview.app.ui.drawer_manager import drawer_manager
ModuleNotFoundError: No module named 'mitraview'
```

**Reply by jourdain** [reply]
It should work with ParaView 5.13.

**Reply by mashirafarid** [reply]
Tried it with ParaView 5.13 and it works if I first specify the venv like below:
```shell
$PVPYTHON --venv .venv -m pytest
```
Thanks so much for all the help!

**Reply by mashirafarid** [reply]
Hi, sorry I had another question related to running pytest, how would I go about using flags (such as -rA) with pytest? I've tried various combos of the command below but the flag either gets confused for a ParaView argument or the tests run normally ignoring the flag
```shell
$PVPYTHON --venv .venv -m pytest -rA tests/
```

**Reply by jourdain** [reply]
If you set your PYTHONPATH/LD_LIBRARY_PATH to point to the ParaView part, you can use your plain python/venv to run things.

**Reply by mashirafarid** [reply]
I'm unfamiliar with how to set PYTHONPATH/LD_LIBRARY_PATH to point to the ParaView part, but I've set them as follows:
```
PYTHONPATH=/Applications/ParaView-5.13.0.app/Contents/Python
LD_LIBRARY_PATH=/Applications/ParaView-5.13.0.app/Contents/Libraries/lib
``` 
After that, I tried using this command to run the tests but get this error regarding numpy. 
```shell
% pytest -rA tests/
Traceback:
/Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/__init__.py:23: in <module>
    from . import multiarray
/Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/multiarray.py:10: in <module>
    from . import overrides
/Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/overrides.py:8: in <module>
    from numpy.core._multiarray_umath import (
E   ImportError: dlopen(/Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/_multiarray_umath.cpython-310-darwin.so, 0x0002): Library not loaded: @executable_path/../Libraries/libcblas.3.11.0.dylib
E     Referenced from: <7D317C4B-C4A5-3D14-85E3-76CC30FC0C45> /Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/_multiarray_umath.cpython-310-darwin.so
E     Reason: tried: '/Users/mashirafarid/.pyenv/versions/3.10.14/Libraries/libcblas.3.11.0.dylib' (no such file), '/usr/local/lib/libcblas.3.11.0.dylib' (no such file), '/usr/lib/libcblas.3.11.0.dylib' (no such file, not in dyld cache)

During handling of the above exception, another exception occurred:
../../../../.pyenv/versions/3.10.14/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_pressure_gradient.py:3: in <module>
    import numpy as np
/Applications/ParaView-5.13.0.app/Contents/Python/numpy/__init__.py:139: in <module>
    from . import core
/Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/__init__.py:49: in <module>
    raise ImportError(msg)
E   ImportError: 
E   
E   IMPORTANT: PLEASE READ THIS FOR ADVICE ON HOW TO SOLVE THIS ISSUE!
E   
E   Importing the numpy C-extensions failed. This error can happen for
E   many reasons, often due to issues with your setup or how NumPy was
E   installed.
E   
E   We have compiled some common reasons and troubleshooting tips at:
E   
E       https://numpy.org/devdocs/user/troubleshooting-importerror.html
E   
E   Please note and check the following:
E   
E     * The Python version is: Python3.10 from "/Users/mashirafarid/Documents/University things/Thesis/mitraview/.venv/bin/python"
E     * The NumPy version is: "1.25.2"
E   
E   and make sure that they are the versions you expect.
E   Please carefully study the documentation linked above for further help.
E   
E   Original error was: dlopen(/Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/_multiarray_umath.cpython-310-darwin.so, 0x0002): Library not loaded: @executable_path/../Libraries/libcblas.3.11.0.dylib
E     Referenced from: <7D317C4B-C4A5-3D14-85E3-76CC30FC0C45> /Applications/ParaView-5.13.0.app/Contents/Python/numpy/core/_multiarray_umath.cpython-310-darwin.so
E     Reason: tried: '/Users/mashirafarid/.pyenv/versions/3.10.14/Libraries/libcblas.3.11.0.dylib' (no such file), '/usr/local/lib/libcblas.3.11.0.dylib' (no such file), '/usr/lib/libcblas.3.11.0.dylib' (no such file, not in dyld cache)
``` 
I had a more recent version of numpy installed in my venv so I tried uninstalling it and installing 1.25.2 instead but that hasn't changed anything. I also read through the page linked in the error but nothing seems related to my error. How can I fix this?

**Reply by jourdain** [reply]
Remove numpy from your venv as it is already in ParaView and conflicting with yours.

**Reply by mashirafarid** [reply]
I've removed it both from my system and from the venv, but still get the same error

**Reply by jourdain** [reply]
Is your venv created from a python 3.10?

**Reply by mashirafarid** [reply]
Yes, these are the commands I use to set it up:
```shell
python3.10 -m venv .venv
source .venv/bin/activate
python -m pip install -U pip
pip install -e .
deactivate
``` 
In my `setup.cfg` file I have these listed as required packages:
```
install_requires =
    trame
    trame-vuetify
    trame-vtk
    fpdf2
    pytest
    scipy
    numpy
    pytest
```

**Reply by jourdain** [reply]
Those comes with ParaView and therefore, you should not have those in your venv.

```
scipy
numpy
```

**Reply by mashirafarid** [reply]
After removing those two packages from `setup.cfg` and recreating my venv, I now get a new error
```shell
Traceback:
/Applications/ParaView-5.13.0.app/Contents/Python/paraview/__init__.py:217: in <module>
    from .modules import vtkRemotingCore
E   ImportError: dlopen(/Applications/ParaView-5.13.0.app/Contents/Python/paraview/modules/vtkRemotingCore.so, 0x0002): Library not loaded: @executable_path/../Libraries/libvtkRemotingCore-pv5.13.5.13.dylib
E     Referenced from: <2716443C-0217-36FF-882A-67D7B5309D9B> /Applications/ParaView-5.13.0.app/Contents/Python/paraview/modules/vtkRemotingCore.so
E     Reason: tried: '/Users/mashirafarid/.pyenv/versions/3.10.14/Libraries/libvtkRemotingCore-pv5.13.5.13.dylib' (no such file), '/usr/local/lib/libvtkRemotingCore-pv5.13.5.13.dylib' (no such file), '/usr/lib/libvtkRemotingCore-pv5.13.5.13.dylib' (no such file, not in dyld cache)

During handling of the above exception, another exception occurred:
../../../../.pyenv/versions/3.10.14/lib/python3.10/importlib/__init__.py:126: in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
tests/test_min_LVOT_area.py:4: in <module>
    from paraview import simple
/Applications/ParaView-5.13.0.app/Contents/Python/paraview/__init__.py:219: in <module>
    import _paraview_modules_static
E   ModuleNotFoundError: No module named '_paraview_modules_static'
```

**Reply by jourdain** [reply]
Since you are on mac and not on linux you should use DYLD_LIBRARY_PATH instead of LD_LIBRARY_PATH as the ENV variable name.

**Reply by mashirafarid** [reply]
I've tried using DYLD_LIBRARY_PATH instead but when I check what env variables are set it isn't there

**Comment by jourdain** [comment]
On mac and linux, you need to use `export DYLD_LIBRARY_PATH=...` not `set`.
